File: frontend/src/types/billing.ts
export interface PlanPriceOut {
  id: number
  plan_id: number
  amount: number
  currency: string
  interval: 'month' | 'year'
  interval_count: number
  external_price_id: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface PlanOut {
  id: number
  name: string
  description: string | null
  external_product_id: string | null
  is_active: boolean
  prices: PlanPriceOut[]
  created_at: string
  updated_at: string
}

export interface SubscriptionOut {
  id: number
  tenant_id: number
  plan_price_id: number | null
  external_subscription_id: string | null
  external_customer_id: string | null
  status: 'incomplete' | 'active' | 'trialing' | 'past_due' | 'canceled'
  current_period_start: string | null
  current_period_end: string | null
  cancel_at_period_end: boolean
  canceled_at: string | null
  plan_price: PlanPriceOut | null
  created_at: string
  updated_at: string
}

export interface CheckoutOut {
  checkout_url: string
  external_session_id: string
}

export interface CustomerPortalOut {
  portal_url: string
}

export type SwitchPlanOut = SubscriptionOut | CheckoutOut

File: backend/src/schemas/billing.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, computed_field

from src.core.config import settings


class CheckoutIn(BaseModel):
    plan_price_id: int


class SwitchPlanIn(BaseModel):
    plan_price_id: int


class StartTrialIn(BaseModel):
    plan_price_id: int


class PlanPriceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan_id: int
    amount: int
    currency: str
    interval: str
    interval_count: int
    is_active: bool

    @computed_field  # type: ignore[prop-decorator]
    @property
    def trial_period_days(self) -> int | None:
        return settings.billing_trial_period_days or None

    created_at: datetime
    updated_at: datetime


class PlanOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    is_active: bool
    prices: list[PlanPriceOut]
    created_at: datetime
    updated_at: datetime


class SubscriptionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    organization_id: int
    plan_price_id: int | None
    status: str
    current_period_start: datetime | None
    current_period_end: datetime | None
    cancel_at_period_end: bool
    canceled_at: datetime | None
    cancel_at: datetime | None
    trial_end: datetime | None
    plan_price: PlanPriceOut | None
    has_payment_method: bool = False
    trial_used: bool = False
    created_at: datetime
    updated_at: datetime


class CheckoutOut(BaseModel):
    checkout_url: str


class CustomerPortalOut(BaseModel):
    portal_url: str
